rust/kernel/delay.rs

```diff
@@ -27,6 +21,52 @@ pub fn coarse_sleep(duration: Duration) {
         u64::from(core::ffi::c_uint::MAX),
     ) as core::ffi::c_uint;
 
+    seconds_as_millis.saturating_add(nanos_as_millis)
+}
+
+/// Sleeps safely even with waitqueue interruptions.
+///
+/// This function forwards the call to the C side `msleep` function. As a result,
+/// `duration` will be rounded up to the nearest millisecond if granularity less
+/// than a millisecond is provided. Any [`Duration`] that exceeds
+/// [`c_uint::MAX`][core::ffi::c_uint::MAX] in milliseconds is saturated.
+///
+/// # Examples
+///
+/// ```
+/// # use core::time::Duration;
+/// # use kernel::delay::coarse_sleep;
+/// coarse_sleep(Duration::new(0, 0));          // Equivalent to `msleep(0)`.
```

Comment:
These may be easier to read if written:
```suggestion
/// coarse_sleep(Duration::from_secs(0));          // Equivalent to `msleep(0)`.
```

Reply:
I considered that, as well as using `ZERO` etc., but I wanted to keep all the examples the same so that it was easy to tell the "jump" from one ns to the next as well as from one milli to the next (in the `msleep` equivalent), if that makes sense.

Also, given there are several examples plus the equivalence in milliseconds in the comment, it should be fairly clear what the arguments are (i.e. seconds and nanoseconds).

Reply:
On the other hand, people may copy-paste this and not know there are better constructors, so perhaps it is worth doing it (but keeping the `#[test]`s with `new` for the reasons above).

Reply:
After thinking about it, I agree it is best to show the friendlier constructors if only because those are what people should likely be using in actual driver code etc. And with a few changes in the examples the "jump"s in milliseconds are still obvious.

Thanks @alex!